### ISSUE TITLE
Error assigning primary key

### ISSUE THREAD

**Issue by twsl** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1484]
### Steps to reproduce
I am trying to scaffold an old database with the following command: `dotnet ef dbcontext scaffold "Server=192.168.1.11;Database=<db>;User=root;Password=<pw>;" "Pomelo.EntityFrameworkCore.MySql" -v -f --prefix-output`

### The issue
Scaffolding doesn't work.

If you are seeing an exception, include the full exceptions details (message and stack trace).

```
verbose: No explicit ServerVersion was set.
verbose: ServerVersion '5.1.36-mysql' was automatically detected.
info:    Using ServerVersion '5.1.36-mysql'.
error:   Error assigning primary key for tbl1.
error:   Error assigning primary key for tbl2.
error:   Error assigning primary key for tbl3.
error:   Error assigning index for tbl1.
error:   Error assigning index for tbl1.
error:   Error assigning index for tbl2.
error:   Error assigning index for tbl3.
warn:    Could not scaffold the foreign key 'tbl1(A4Id)'. A key for 'A4Id' was not found in the principal entity type 'Tbl4'.
warn:    Could not scaffold the foreign key 'tbl2(A1Id)'. A key for 'A1Id' was not found in the principal entity type 'Tbl1'.
warn:    Could not scaffold the foreign key 'tbl3(T2Id)'. A key for 'T2Id' was not found in the principal entity type 'Tbl2'.
```
I modified the property and table names for privacy reasons.

### Further technical details

MySQL version: 5.1.36
Operating system: Windows
Pomelo.EntityFrameworkCore.MySql version: 5.0.0
Microsoft.AspNetCore.App version: 5

Other details about my project setup: The weird thing is, that I do have a second database running `5.5.57-MariaDB` where the a database with a different version of the same structure can be found and in phpMyAdmin the configuration looks identical. MariaDb works, MySql doesn't. According to phpMyAdmin all keys are set corretly in both dbs.

**Comment by twsl** [comment]
Simply trying to copy the configuration from the other database results in:
`MySqlConnector.MySqlException (0x80004005): Incorrect key file for table '/tmp/#sql_28ce_0.MYI'; try to repair it`.
Tablecheck in phpMyAdmin says everything is fine.

**Comment by mguinness** [comment]
Referring to [Supported Databases and Versions](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#supported-databases-and-versions) that version of MySQL isn't supported.

If you want to dig further, you could look at the queries against `INFORMATION_SCHEMA` in  [MySqlDatabaseModelFactory](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/EFCore.MySql/Scaffolding/Internal/MySqlDatabaseModelFactory.cs) and compare the output between MySQL and MariaDB.

For your second issue, take a look at [MySQL “Incorrect key file for table /tmp/#sql_xxx_x.MYI” error](https://electrictoolbox.com/mysql-incorrect-key-file-for-table-tmp-sql-myi/).

**Comment by twsl** [comment]
Thanks @mguinness for showing me the direction.
In MySql 5.1 the `Sub_PARTS` from the [GetPrimaryQuery](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/EFCore.MySql/Scaffolding/Internal/MySqlDatabaseModelFactory.cs#L562) is a blob rather than a string as it is in MariaDb.
How to we handle this properly?

**Comment by twsl** [comment]
I still haven't verified this by debugging, but it looks like the parsing of the `SUB_PARTS` is something that could need some additional handling instead of just reading a string.
```
If group_concat_max_len <= 512, the return type is VARBINARY or VARCHAR; otherwise, the return type is BLOB or TEXT.
The choice between binary or non-binary types depends from the input.
```
[source](https://mariadb.com/kb/en/group_concat/)

```
The return value is a nonbinary or binary string, depending on whether the arguments are nonbinary or binary strings.
The result type is TEXT or BLOB unless group_concat_max_len is less than or equal to 512, in which case the result type is VARCHAR or VARBINARY.
```
[source](https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions.html#function_group-concat)

**Comment by mguinness** [comment]
What is the current value of `group_concat_max_len` in both your MySQL and MariaDB instances?  How many columns are in the  indexes concerned on tbl1, tbl2 & tbl3?

**Comment by twsl** [comment]
`group_concat_max_len` is 1024 in both cases. All tables have less than 50 columns.
As far as I understand the documentation, there is no way to know, whether you receive a blob or text as return type.

**Comment by mguinness** [comment]
I'm curious how many of those columns you actually have in the index.  Even w/ 512 bytes that means up to 128 characters (utf8mb4).

**Comment by twsl** [comment]
Not too sure, if I understand what you'd like to know, but the result of all `GetPrimarQueries` looks like this:
MySql:
|INDEX_NAME |COLUMNS |SUB_PARTS |
|---|---|---|
|PRIMARY |A1Id| [BLOB - 1Bytes] |

MariaDb:
|INDEX_NAME |COLUMNS |SUB_PARTS |
|---|---|---|
|PRIMARY |A1Id| 0 |

**Comment by mguinness** [comment]
I did some research and this only occurs in version 5.1 of MySQL:

> Implicit conversion of a number or temporal value to a string produces a binary string (a `BINARY`, `VARBINARY`, or `BLOB` value). Such implicit conversions to string typically occur for functions that are passed numeric values when string values are more usual, and thus can have effects beyond the type of the converted value. Consider the expression `CONCAT(1, 'abc')`. The numeric argument 1 is converted to the binary string '1' and the concatenation of that value with the nonbinary string 'abc' produces the binary string '1abc'.

In MySQL 5.5 release notes where this behavior was corrected:

> Previously, an implicit conversion always produced a binary string, regardless of the connection character set. Such implicit conversions to string typically occur for functions that are passed numeric or temporal values when string values are more usual, and thus could have effects beyond the type of the converted value.

A workaround for 5.1 is to change the function parameter by casting the value to char beforehand:

`` GROUP_CONCAT(CAST(IFNULL(`SUB_PART`, 0) AS CHAR) ORDER BY `SEQ_IN_INDEX` SEPARATOR ',') AS `SUB_PARTS` ``

The question is do we want to cater to a database release where official support ended in Dec 2013?

**Comment by twsl** [comment]
Wow, good catch!
I'd love the see such a fix for two reasons: primarily to document the weird quirks of MySql and secondly that would probably the easiest solution for me personally.
However, I can fully understand, if fixing a bug for a version that went out of support years ago and is not on the list of supported versions is out of scope.
But, as far as I understand the docs, there is still no way to know whether the return type is always a string. A generic fix for BLOB return type would be my proposed way to solve this.

**Comment by mguinness** [comment]
My understanding is that `GROUP_CONCAT` will only return `BLOB` with binary input.  Based on the fix [WL#2649: Number-to-string conversions](https://dev.mysql.com/worklog/task/?id=2649) in MySQL 5.5, numeric values will be nonbinary and will return `TEXT`.

If you need a fix for 5.1, you could implement the workaround I suggested as a PR.  To enable it only for certain versions you would have to add a flag to both [MySqlServerVersion](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/97898733f89430ab0cfaa0209ab716a1f7d664f1/src/EFCore.MySql/Infrastructure/MySqlServerVersion.cs) and [MariaDbServerVersion](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/97898733f89430ab0cfaa0209ab716a1f7d664f1/src/EFCore.MySql/Infrastructure/MariaDbServerVersion.cs) classes and then use `_options.ServerVersion.Supports`.

**Comment by twsl** [comment]
I created a draft PR, but the tests fail and I cannot see the azure pipelines error.
